zqd/ingest/archivepcap.go

```diff
@@ -0,0 +1,247 @@
+	group, ctx := errgroup.WithContext(ctx)
+	var pipes []*io.PipeWriter
+	var zreaders []zbuf.Reader
+	if p.zeek != nil {
+		pw, dr, err := p.runAnalyzer(ctx, group, p.zeek)
+		if err != nil {
+			return nil, nil, err
+		}
+		pipes = append(pipes, pw)
+		zreaders = append(zreaders, dr)
+	}
+	if p.suricata != nil {
+		pw, dr, err := p.runAnalyzer(ctx, group, p.suricata)
+		if err != nil {
+			return nil, nil, err
+		}
+		pipes = append(pipes, pw)
+		// Suricata logs need flowgraph to rename timestamp fields into ts.
+		tr, err := driver.NewReader(ctx, suricataTransform, p.zctx, dr)
+		if err != nil {
+			return nil, nil, err
+		}
+		zreaders = append(zreaders, tr)
+	}
+	group.Go(func() error {
+		var writers []io.Writer
+		for _, p := range pipes {
+			writers = append(writers, p)
+		}
+		_, err := ctxio.Copy(ctx, io.MultiWriter(writers...), pcapstream)
+		// Once copy has completed, close pipe writers which will instruct the
+		// analyzer processes to exit.
+		for _, p := range pipes {
+			p.Close()
+		}
+		return err
+	})
+	return group, zreaders, nil
+}
+
+func (p *archivePcapOp) runAnalyzer(ctx context.Context, group *errgroup.Group, ln pcapanalyzer.Launcher) (*io.PipeWriter, zbuf.Reader, error) {
+	logdir, err := ioutil.TempDir("", "zqd-pcap-ingest-")
+	if err != nil {
+		return nil, nil, err
+	}
+	p.cleanup(func() { os.RemoveAll(logdir) })
+	pr, pw := io.Pipe()
+	waiter, err := ln(ctx, pr, logdir)
+	if err != nil {
+		return nil, nil, err
+	}
+	dr, err := newLogTailer(p.zctx, logdir, zio.ReaderOpts{
+		JSON: ndjsonio.ReaderOpts{TypeConfig: suricataTC},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	group.Go(func() error {
+		err := waiter.Wait()
+		// Analyzer has either encountered an error or recieved an EOF from the
+		// pcap stream. Tell DirReader to stop tail files, which will in turn
+		// cause an EOF on zbuf.Read stream when remaining data has been read.
+		if errs := dr.Stop(); err == nil {
+			err = errs
+		}
+		return err
+	})
+	return pw, dr, nil
+}
+
+func (p *archivePcapOp) Status() api.PcapPostStatus {
+	return api.PcapPostStatus{
+		Type:         "PcapPostStatus",
+		StartTime:    p.startTime,
+		UpdateTime:   nano.Now(),
+		PcapSize:     p.pcapBytesTotal,
+		PcapReadSize: atomic.LoadInt64(&p.pcapBytesRead),
+		RecordCount:  p.recordCounter.Records(),
+		RecordBytes:  p.recordCounter.Bytes(),
+	}
+}
+
+type recordCounter struct {
+	reader      zbuf.Reader
+	bytesRead   int64
+	recordsRead int64
+}
+
+func (r *recordCounter) Read() (*zng.Record, error) {
+	rec, err := r.reader.Read()
+	if rec != nil {
+		atomic.AddInt64(&r.bytesRead, int64(len(rec.Raw)))
+		atomic.AddInt64(&r.recordsRead, 1)
+	}
+	return rec, err
+}
+
+func (r *recordCounter) Bytes() int64 {
+	return atomic.LoadInt64(&r.bytesRead)
+}
+
+func (r *recordCounter) Records() int64 {
+	return atomic.LoadInt64(&r.recordsRead)
+}
+
+// Snap for archivePcapOp is functionally useless. It is only here to satisfy
```

Comment:
When we can run Brim against this branch, I'm curious how this impacts the live updating. I thought Brim triggered updating its view during pcap by looking for new snapshots, but I haven't looked at that code in while.

Reply:
As discussed offline, it is. Without an incrementing snapshot brim will not know new data is available to query. As discussed, this should really be keeping track of record count (and bytes) written to the archive, return it in the status updates, and Brim should use this to establish if new data is available.

Reply:
#1488